src/server_manager/electron_app/digitalocean_oauth.ts

```diff
@@ -0,0 +1,216 @@
+  server.on('close', () => console.log('Oauth server closed'));
+
+  let isCancelled = false;
+  app.use((req, res, next) => {
+    if (isCancelled) {
+      res.status(503).send('Authentication cancelled');
+    } else {
+      next();
+    }
+  });
+
+  app.get('/', (request, response) => {
+    response.send(`<html>
+          <head><title>Authenticating...</title></head>
+          <body>
+              <form id="form" method="POST">
+                  <input id="params" type="hidden" name="params"></input>
+              </form>
+              <script>
+                  let params = new URLSearchParams(location.hash.substr(1));
+                  let form = document.getElementById("form");
+                  let targetUrl = params.get("state");
+                  form.setAttribute("action", targetUrl);
+                  document.getElementById("params").setAttribute("value", params);
+                  form.submit();
+              </script>
+          </body>
+      </html>`);
+  });
+
+  const rejectWrapper = {reject: (error: Error) => {}};
+  const result = new Promise<string>((resolve, reject) => {
+    rejectWrapper.reject = reject;
+    app.post('/', bodyParser.urlencoded({type: '*/*', extended: false}), (request, response) => {
+      server.close();
+
+      const requestSecret = request.query.secret;
+      if (requestSecret !== secret) {
+        response.status(400).send('Authentication failed');
+        reject(new Error(`Expected secret ${secret}. Got ${requestSecret}`));
+        return;
+      }
+      const params = new URLSearchParams(request.body.params);
+      if (params.get('error')) {
+        response.status(400).send('Authentication failed');
+        reject(new Error(`DigitalOcean OAuth error: ${params.get('error_description')}`));
+        return;
+      }
+      const accessToken = params.get('access_token');
+      if (accessToken) {
+        getAccount(accessToken)
+            .then((account) => {
+              if (account.status === 'active') {
+                response.send(`<html><script>window.close()</script></html>`);
+              } else {
+                response.redirect('https://cloud.digitalocean.com');
+              }
+              resolve(accessToken);
+            })
+            .catch(reject);
+      } else {
+        response.status(400).send('Authentication failed');
+        reject(new Error('No access_token on OAuth response'));
+      }
+    });
+
+    listenOnFirstPort(server, REGISTERED_REDIRECTS.map(e => e.port))
+        .then((index) => {
+          const port = REGISTERED_REDIRECTS[index].port;
+          const clientId = REGISTERED_REDIRECTS[index].clientId;
+          const address = server.address();
+          console.log(`OAuth target listening on ${address.address}:${address.port}`);
+
+          const targetUrl = `http://localhost:${
+              encodeURIComponent(address.port.toString())}?secret=${encodeURIComponent(secret)}`;
```

Comment:
These calls to `encodeURIComponent` are both unnecessary, right?

Reply:
I'm being excessive here because they are provided by a different piece of code.